### PR TITLE
chore: Check request headers for `:status`

### LINKED ISSUE #1274: check that - for a request, we DON'T get :status?
I know that this is old code, but shouldn't we be checking that - for a request, we DON'T get :status?

That means we need a set of bits that should be on and a set of bits that should be off.  Or:

```rust
        const PSEUDO_HEADER_ALL: u8 = PSEUDO_HEADER_STATUS | PSEUDO_HEADER_METHOD | PSEUDO_HEADER_SCHEME | PSEUDO_HEADER_AUTHORITY | PSEUDO_HEADER_PATH;
        let (ph_required, ph_mask) = match message_type {
             // Responses contain only :status
             MessageType::Response => (PSEUDO_HEADER_STATUS, PSEUDO_HEADER_ALL),
            MessageType::Request => {
                if method_value == Some(&"CONNECT".to_string()) {
                    // CONNECT requests contain only :method and :authority
                    // This won't work once we get to :protocol...
                    (PSEUDO_HEADER_METHOD | PSEUDO_HEADER_AUTHORITY, PSEUDO_HEADER_ALL)
                } else {
                    // Other requests contain :method, :scheme, and :path, and may contain :authority.
                    (PSEUDO_HEADER_METHOD | PSEUDO_HEADER_SCHEME | PSEUDO_HEADER_PATH, PSEUDO_HEADER_ALL & !PSEUDO_HEADER_AUTHORITY)
                }
            }
        };
        if pseudo_state & ph_mask != ph_required {
            return Err(Error::InvalidHeader);
        }
```

_Originally posted by @martinthomson in https://github.com/mozilla/neqo/pull/1270#discussion_r755542245_

### PR DESCRIPTION
**Pull request by larseggert**
This tries to implement the suggestion @martinthomson made in #1274. Needs a review from someone who knows the code better than me.

Fixes #1274

### CODE REVIEW

File: neqo-http3/src/headers_checks.rs [hunk 1]
```diff
@@ -119,24 +126,32 @@ pub fn headers_valid(headers: &[Header], message_type: MessageType) -> Res<()> {
     }
     // Clear the regular header bit, since we only check pseudo headers below.
     pseudo_state.remove(PseudoHeaderState::Regular);
-    let pseudo_header_mask = match message_type {
-        MessageType::Response => enum_set!(PseudoHeaderState::Status),
+    let (pseudo_header_required, pseudo_header_mask) = match message_type {
+        // Responses contain only :status
+        MessageType::Response => (enum_set!(PseudoHeaderState::Status), PSEUDO_HEADER_ALL),
         MessageType::Request => {
             if method_value == Some("CONNECT") {
```
Comment (mxinden) [inline neqo-http3/src/headers_checks.rs:133]: Feel free to ignore.

```suggestion
        MessageType::Request if method_value == Some("CONNECT") => {
```

Given that the `if` `else` block spans the entire `match` arm, the above would safe a level of indentation.

I can also propose this as a follow-up.

**Review by martinthomson — CHANGES_REQUESTED** [review]
This looks fine, but it really needs a test to prove that it has addressed the issue.

File: neqo-http3/src/headers_checks.rs (hunk 1 above)
Comment (larseggert) [reply neqo-http3/src/headers_checks.rs]: `rustfmt` seems to want to add the braces back around the `if`.

**Comment by larseggert** [comment]
So I added a test, and it passes even without this PR. I think this is because @KershawChang already fixed this back in #1022?

File: neqo-http3/src/headers_checks.rs (hunk 1 above)
Comment (martinthomson) [reply neqo-http3/src/headers_checks.rs]: So you can't do:

```rust
match whatever {
  MessageType::Request if method_value == Some("CONNECT") => { code }
  MessageType::Request => { else_clause }
}
```
?

**Comment by larseggert** [comment]
@KershawChang do you recall fixing this? If yes, I will happily close this PR.

**Comment by KershawChang** [comment]
> @KershawChang do you recall fixing this? If yes, I will happily close this PR.

Sorry for the delayed reply.
Based on the code below, yes, I think this was fixed.
https://github.com/mozilla/neqo/blob/ad8f3909d9ae96219e56a7c31e6822d6e64a6dc9/neqo-http3/src/headers_checks.rs#L138

I think we can close this PR. Thanks.

**Pull request closed without merging**